server/src/test/java/org/elasticsearch/cluster/ClusterStateTests.java

```diff
@@ -1171,4 +1171,74 @@ public void testGetMinTransportVersion() throws IOException {
         );
     }
+
+    public static int expectedChunkCount(ToXContent.Params params, ClusterState clusterState) {
+        final var metrics = ClusterState.Metric.parseString(params.param("metric", "_all"), true);
+
+        int chunkCount = 0;
+
+        // header chunk
+        chunkCount += 1;
+
+        // blocks
+        if (metrics.contains(ClusterState.Metric.BLOCKS)) {
+            chunkCount += 2 + clusterState.blocks().indices().size();
+        }
+
+        // nodes, transport_versions, nodes_versions
+        if (metrics.contains(ClusterState.Metric.NODES)) {
+            chunkCount += 6 + clusterState.nodes().size() + 2 * clusterState.compatibilityVersions().size();
+        }
+
+        // metadata
+        if (metrics.contains(ClusterState.Metric.METADATA)) {
+            chunkCount += MetadataTests.expectedChunkCount(params, clusterState.metadata());
+        }
+
+        // routing table
+        if (metrics.contains(ClusterState.Metric.ROUTING_TABLE)) {
+            chunkCount += 2;
+            for (IndexRoutingTable indexRoutingTable : clusterState.routingTable()) {
+                chunkCount += 2;
+                for (int shardId = 0; shardId < indexRoutingTable.size(); shardId++) {
+                    chunkCount += 2 + indexRoutingTable.shard(shardId).size();
+                }
+            }
+        }
+
+        // routing nodes
+        if (metrics.contains(ClusterState.Metric.ROUTING_NODES)) {
+            final var routingNodes = clusterState.getRoutingNodes();
+            chunkCount += 4 + routingNodes.unassigned().size();
+            for (RoutingNode routingNode : routingNodes) {
+                chunkCount += 2 + routingNode.size();
+            }
+        }
+
+        // customs
+        if (metrics.contains(ClusterState.Metric.CUSTOMS)) {
+            for (ClusterState.Custom custom : clusterState.customs().values()) {
+                chunkCount += 2;
+
+                if (custom instanceof HealthMetadata) {
+                    chunkCount += 1;
+                } else if (custom instanceof RepositoryCleanupInProgress repositoryCleanupInProgress) {
+                    chunkCount += 2 + repositoryCleanupInProgress.entries().size();
+                } else if (custom instanceof RestoreInProgress restoreInProgress) {
+                    chunkCount += 2 + Iterables.size(restoreInProgress);
+                } else if (custom instanceof SnapshotDeletionsInProgress snapshotDeletionsInProgress) {
+                    chunkCount += 2 + snapshotDeletionsInProgress.getEntries().size();
+                } else if (custom instanceof SnapshotsInProgress snapshotsInProgress) {
+                    chunkCount += 2 + snapshotsInProgress.asStream().count();
+                } else {
+                    // could be anything, we have to just try it
+                    chunkCount += Iterables.size(
+                        (Iterable<ToXContent>) (() -> Iterators.map(custom.toXContentChunked(params), Function.identity()))
```

Comment:
Is the `Iterators.map` needed here?

Reply:
Yeah or at least I couldn't work out a nicer way to satisfy the typechecker that an `Iterator<? extends ToXContent>` is a cromulent `Iterator<ToXContent>`.